src/lib/api.ts

```diff
@@ -27,6 +35,134 @@ export async function createGroup(groupFormValues: GroupFormValues) {
   })
 }
 
+export async function acquireLockRecurringTransaction(groupId:string, expenseId:string, lockId: string): Promise<RecurringTransactions[]> {
```

Comment:
Could you elaborate on why we need to implement locking and retrying here? What is the issue if multiple users edit a recurring expense concurrently? Some comments would go a long way towards explaining the logic.

Reply:
When someone fetches group expenses, only then I'm checking if there's a pending recurring transaction. If yes, then I insert in the DB as new transaction.

Locking is necessary to avoid race condition:
if two people fetch expenses at the same time, APIs of both will result in fetching of pending transactions and inserting in DB. Hence Double insertions for the pending transaction.

Same condition will be there for update. One person is updating the transaction and other is inserting in DB.

Locking ensures consistency and correctness.


Retrying is there to retry acquiring locks after some time if it was acquired by someone else at that instant.

Reply:
Is it possible to delegate this concern to the database? Suppose each expense ID in a recurrence followed the pattern `ffffff01`, `ffffff02`, `ffffff03`, .... When expenses are fetched, you determine the need to create expense `ffffff04`. You do some kind of `INSERT IF NOT EXISTS` query. If there is a race and this expense already got inserted by another request, it doesn't really affect anything.

Worst case scenario I think is if, before your request completes, the expense has already been created *and* then subsequently updated. You might have a stale view of the totals. But it would only be a temporary issue.

Reply:
Umm, not.
Let's take an example:
- 1 Jan 2024 (monthly txn) `ffffff01` was added. And the group wasn't opened until now.
If Person P1 and P2 both open it at the same moment - both will determine that they need to insert 3 new Txns. How will I verify if P1 is already adding the transaction and P2 doesn't need to?
`ffffff04` can be a copy of `ffffff02` (determined by P1) or `ffffff02` (determined by P2).
For realising that someone else is working on this resource, I need to insert a reference id in DB.
That Id can be in the same Expense table, or a new table that I created.
New table made it easy to manage it as I can keep track which txn needs to be inserted without looking into Expense table + separation of concern.

So if lock is not there, in worst case both will insert the same records.

Another case:
P1 didn't call fetch groups (consider he didn't close tab for 4 months). P2 fetches now and he realizes he has to insert 3 new records.

P1 edits the amount from 100 to 200.
What shall happen now? Either P1 should successfully edit first and P2 shall recursively add 3 txns with updated amount OR P2 should insert all 3 txns and then only P1 should update the amount.
It shouldn't happen that P2 inserted 1 record with old amount and later 2 with updated amount.


I agree, chances of happening while edit is very hard, but it's possible that two people open the group at the same time and double inserts can happen.
I made lock so that this doesn't happen.